module/os/linux/spl/spl-thread.c

```diff
@@ -158,3 +158,37 @@ spl_kthread_create(int (*func)(void *), void *data, const char namefmt[], ...)
 	} while (1);
 }
 EXPORT_SYMBOL(spl_kthread_create);
+
+/*
+ * The "why" argument indicates the allowable side-effects of the call:
+ *
+ * FORREAL:  Extract the next pending signal from p_sig into p_cursig;
+ * stop the process if a stop has been requested or if a traced signal
+ * is pending.
+ *
+ * JUSTLOOKING:  Don't stop the process, just indicate whether or not
+ * a signal might be pending (FORREAL is needed to tell for sure).
+ */
+int
+issig(int why)
+{
+	ASSERT(why == FORREAL || why == JUSTLOOKING);
+
+	if (!signal_pending(current))
+		return (0);
+
+	if (why == FORREAL &&
+	    (sigismember(&current->signal->shared_pending.signal, SIGSTOP) ||
+	    sigismember(&current->signal->shared_pending.signal, SIGTSTP))) {
+		/* TODO: right semantics? */
+		int sig = kernel_dequeue_signal();
+		if (sig != SIGSTOP && sig != SIGTSTP)
+			return (1);
```

Comment:
Would this happen if multiple signals are pending? (I'm guessing there's a defined priority of which is dequeued?)  And/or if there's a race condition and a new signal becomes pending after our sigismember() checks?

If that happens and we dequeue'd a different signal, how does that change the behavior of the system?  ZFS will return EINTR but does the kernel still take the intended action before returning from the syscall?

Reply:
I haven't tested with multiple pending signals at once... I believe that it's a queue, so if another signal comes in after we check but before we dequeue, we won't get that signal instead. If we dequeue a signal besides stop, the system still does the right thing; interrupting a send with ^C (or kill -s 2) still results in the process dying, rather than saying that it received EINTR.

Reply:
That matches my understanding as well.

Reply:
It looks like there can be multiple signals pending, and sigismember() just checks if SIGSTOP/SIGTSTP is one of them.  In that case we dequeue a signal, and it’s going to be the one with the lowest value (per https://elixir.bootlin.com/linux/v5.3/source/arch/x86/include/uapi/asm/signal.h#L30), as long as it is not in task->blocked.  so e.g. if SIGUSR1 and SIGSTOP are pending, we will dequeue SIGUSR1.  So it is not a “queue” in the sense that signals are not dequeued in the order they are delivered; they’re delivered in the statically-defined priority order.  My concern is that since we could dequeue a signal other than the ones we care about, how does that signal get delivered?

Reply:
Hopefully this new approach resolves the issue, we use the signal mask to only dequeue STOP and TSTP signals. There is still one hiccup that I don't know what is causing it: if we deliver a signal along with SIGSTOP, usually that signal is delivered after we send SIGCONT to resume the process. However, sometimes that signal doesn't get delivered until we exit kernel space or we send another STOP/CONT pair. I believe, but am not certain, that this has to do with how the SIGPENDING flag works and when it is set and unset.